--- ../src-base/minecraft/net/minecraft/client/renderer/culling/Frustum.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/culling/Frustum.java
@@ -1,10 +1,7 @@
 package net.minecraft.client.renderer.culling;
 
 import net.minecraft.util.AxisAlignedBB;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public class Frustum implements ICamera
 {
     private ClippingHelper field_78552_a;
@@ -29,13 +26,24 @@
         this.field_78549_d = p_78547_5_;
     }
 
+    /**
+     * Calls the clipping helper. Returns true if the box is inside all 6 clipping planes, otherwise returns false.
+     */
     public boolean func_78548_b(double p_78548_1_, double p_78548_3_, double p_78548_5_, double p_78548_7_, double p_78548_9_, double p_78548_11_)
     {
         return this.field_78552_a.func_78553_b(p_78548_1_ - this.field_78550_b, p_78548_3_ - this.field_78551_c, p_78548_5_ - this.field_78549_d, p_78548_7_ - this.field_78550_b, p_78548_9_ - this.field_78551_c, p_78548_11_ - this.field_78549_d);
     }
 
+    /**
+     * Returns true if the bounding box is inside all 6 clipping planes, otherwise returns false.
+     */
     public boolean func_78546_a(AxisAlignedBB p_78546_1_)
     {
         return this.func_78548_b(p_78546_1_.field_72340_a, p_78546_1_.field_72338_b, p_78546_1_.field_72339_c, p_78546_1_.field_72336_d, p_78546_1_.field_72337_e, p_78546_1_.field_72334_f);
     }
+
+    public boolean isBoxInFrustumFully(double p_isBoxInFrustumFully_1_, double p_isBoxInFrustumFully_3_, double p_isBoxInFrustumFully_5_, double p_isBoxInFrustumFully_7_, double p_isBoxInFrustumFully_9_, double p_isBoxInFrustumFully_11_)
+    {
+        return this.field_78552_a.isBoxInFrustumFully(p_isBoxInFrustumFully_1_ - this.field_78550_b, p_isBoxInFrustumFully_3_ - this.field_78551_c, p_isBoxInFrustumFully_5_ - this.field_78549_d, p_isBoxInFrustumFully_7_ - this.field_78550_b, p_isBoxInFrustumFully_9_ - this.field_78551_c, p_isBoxInFrustumFully_11_ - this.field_78549_d);
+    }
 }
