--- ../src-base/minecraft/net/minecraft/client/shader/ShaderUniform.java
+++ ../src-work/minecraft/net/minecraft/client/shader/ShaderUniform.java
@@ -3,14 +3,11 @@
 import java.nio.FloatBuffer;
 import java.nio.IntBuffer;
 import net.minecraft.client.renderer.OpenGlHelper;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.lwjgl.BufferUtils;
 import org.lwjgl.util.vector.Matrix4f;
 
-@SideOnly(Side.CLIENT)
 public class ShaderUniform
 {
     private static final Logger field_148104_a = LogManager.getLogger();
@@ -263,15 +260,19 @@
             case 0:
                 OpenGlHelper.func_153181_a(this.field_148102_b, this.field_148101_e);
                 break;
+
             case 1:
                 OpenGlHelper.func_153182_b(this.field_148102_b, this.field_148101_e);
                 break;
+
             case 2:
                 OpenGlHelper.func_153192_c(this.field_148102_b, this.field_148101_e);
                 break;
+
             case 3:
                 OpenGlHelper.func_153162_d(this.field_148102_b, this.field_148101_e);
                 break;
+
             default:
                 field_148104_a.warn("Uniform.upload called, but count value (" + this.field_148103_c + ") is " + " not in the range of 1 to 4. Ignoring.");
         }
@@ -284,15 +285,19 @@
             case 4:
                 OpenGlHelper.func_153168_a(this.field_148102_b, this.field_148098_f);
                 break;
+
             case 5:
                 OpenGlHelper.func_153177_b(this.field_148102_b, this.field_148098_f);
                 break;
+
             case 6:
                 OpenGlHelper.func_153191_c(this.field_148102_b, this.field_148098_f);
                 break;
+
             case 7:
                 OpenGlHelper.func_153159_d(this.field_148102_b, this.field_148098_f);
                 break;
+
             default:
                 field_148104_a.warn("Uniform.upload called, but count value (" + this.field_148103_c + ") is " + "not in the range of 1 to 4. Ignoring.");
         }
@@ -305,9 +310,11 @@
             case 8:
                 OpenGlHelper.func_153173_a(this.field_148102_b, true, this.field_148098_f);
                 break;
+
             case 9:
                 OpenGlHelper.func_153189_b(this.field_148102_b, true, this.field_148098_f);
                 break;
+
             case 10:
                 OpenGlHelper.func_153160_c(this.field_148102_b, true, this.field_148098_f);
         }
