--- ../src-base/minecraft/net/minecraft/client/renderer/RegionRenderCache.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RegionRenderCache.java
@@ -1,8 +1,11 @@
 package net.minecraft.client.renderer;
 
 import java.util.Arrays;
+
+import net.mattbenson.Wrapper;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
+import net.minecraft.src.Config;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.Vec3i;
@@ -11,6 +14,7 @@
 import net.minecraft.world.chunk.Chunk;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import net.optifine.DynamicLights;
 
 @SideOnly(Side.CLIENT)
 public class RegionRenderCache extends ChunkCache
@@ -39,12 +43,23 @@
 
     public int func_175626_b(BlockPos p_175626_1_, int p_175626_2_)
     {
+    	//TODO: Falcun patcher
+    	if(Wrapper.getInstance().isRemoveLightCalculations()) {
+    		return 1000;
+    	}
+    	
         int i = this.func_175630_e(p_175626_1_);
         int j = this.field_175634_h[i];
 
         if (j == -1)
         {
             j = super.func_175626_b(p_175626_1_, p_175626_2_);
+
+            if (Config.isDynamicLights() && !this.func_180495_p(p_175626_1_).func_177230_c().func_149662_c())
+            {
+                j = DynamicLights.getCombinedLight(p_175626_1_, j);
+            }
+
             this.field_175634_h[i] = j;
         }
 
@@ -53,16 +68,23 @@
 
     public IBlockState func_180495_p(BlockPos p_180495_1_)
     {
-        int i = this.func_175630_e(p_180495_1_);
-        IBlockState iblockstate = this.field_175635_i[i];
+    	try {
+            int i = this.func_175630_e(p_180495_1_);
+            IBlockState iblockstate = this.field_175635_i[i];
 
-        if (iblockstate == null)
-        {
-            iblockstate = this.func_175631_c(p_180495_1_);
-            this.field_175635_i[i] = iblockstate;
-        }
+            if (iblockstate == null)
+            {
+                iblockstate = this.func_175631_c(p_180495_1_);
+                this.field_175635_i[i] = iblockstate;
+            }
 
-        return iblockstate;
+            return iblockstate;
+    		}
+    		catch(Exception e) {
+
+    	        return null;
+    		}
+
     }
 
     private IBlockState func_175631_c(BlockPos p_175631_1_)
