--- ../src-base/minecraft/net/minecraft/client/renderer/RenderList.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderList.java
@@ -1,30 +1,101 @@
 package net.minecraft.client.renderer;
 
+import java.nio.IntBuffer;
 import net.minecraft.client.renderer.chunk.ListedRenderChunk;
 import net.minecraft.client.renderer.chunk.RenderChunk;
+import net.minecraft.src.Config;
 import net.minecraft.util.EnumWorldBlockLayer;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 import org.lwjgl.opengl.GL11;
 
-@SideOnly(Side.CLIENT)
 public class RenderList extends ChunkRenderContainer
 {
+    private double viewEntityX;
+    private double viewEntityY;
+    private double viewEntityZ;
+    IntBuffer bufferLists = GLAllocation.func_74527_f(16);
+
     public void func_178001_a(EnumWorldBlockLayer p_178001_1_)
     {
         if (this.field_178007_b)
         {
-            for (RenderChunk renderchunk : this.field_178009_a)
+            if (!Config.isRenderRegions())
             {
-                ListedRenderChunk listedrenderchunk = (ListedRenderChunk)renderchunk;
-                GlStateManager.func_179094_E();
-                this.func_178003_a(renderchunk);
-                GL11.glCallList(listedrenderchunk.func_178600_a(p_178001_1_, listedrenderchunk.func_178571_g()));
-                GlStateManager.func_179121_F();
+                for (RenderChunk renderchunk1 : this.field_178009_a)
+                {
+                    ListedRenderChunk listedrenderchunk1 = (ListedRenderChunk)renderchunk1;
+                    GlStateManager.func_179094_E();
+                    this.func_178003_a(renderchunk1);
+                    GL11.glCallList(listedrenderchunk1.func_178600_a(p_178001_1_, listedrenderchunk1.func_178571_g()));
+                    GlStateManager.func_179121_F();
+                }
             }
+            else
+            {
+                int i = Integer.MIN_VALUE;
+                int j = Integer.MIN_VALUE;
 
+                for (RenderChunk renderchunk : this.field_178009_a)
+                {
+                    ListedRenderChunk listedrenderchunk = (ListedRenderChunk)renderchunk;
+
+                    if (i != renderchunk.regionX || j != renderchunk.regionZ)
+                    {
+                        if (this.bufferLists.position() > 0)
+                        {
+                            this.drawRegion(i, j, this.bufferLists);
+                        }
+
+                        i = renderchunk.regionX;
+                        j = renderchunk.regionZ;
+                    }
+
+                    if (this.bufferLists.position() >= this.bufferLists.capacity())
+                    {
+                        IntBuffer intbuffer = GLAllocation.func_74527_f(this.bufferLists.capacity() * 2);
+                        this.bufferLists.flip();
+                        intbuffer.put(this.bufferLists);
+                        this.bufferLists = intbuffer;
+                    }
+
+                    this.bufferLists.put(listedrenderchunk.func_178600_a(p_178001_1_, listedrenderchunk.func_178571_g()));
+                }
+
+                if (this.bufferLists.position() > 0)
+                {
+                    this.drawRegion(i, j, this.bufferLists);
+                }
+            }
+
+            if (Config.isMultiTexture())
+            {
+                GlStateManager.bindCurrentTexture();
+            }
+
             GlStateManager.func_179117_G();
             this.field_178009_a.clear();
         }
     }
+
+    public void func_178004_a(double viewEntityXIn, double viewEntityYIn, double viewEntityZIn)
+    {
+        this.viewEntityX = viewEntityXIn;
+        this.viewEntityY = viewEntityYIn;
+        this.viewEntityZ = viewEntityZIn;
+        super.func_178004_a(viewEntityXIn, viewEntityYIn, viewEntityZIn);
+    }
+
+    private void drawRegion(int p_drawRegion_1_, int p_drawRegion_2_, IntBuffer p_drawRegion_3_)
+    {
+        GlStateManager.func_179094_E();
+        this.preRenderRegion(p_drawRegion_1_, 0, p_drawRegion_2_);
+        p_drawRegion_3_.flip();
+        GlStateManager.callLists(p_drawRegion_3_);
+        p_drawRegion_3_.clear();
+        GlStateManager.func_179121_F();
+    }
+
+    public void preRenderRegion(int p_preRenderRegion_1_, int p_preRenderRegion_2_, int p_preRenderRegion_3_)
+    {
+        GlStateManager.func_179109_b((float)((double)p_preRenderRegion_1_ - this.viewEntityX), (float)((double)p_preRenderRegion_2_ - this.viewEntityY), (float)((double)p_preRenderRegion_3_ - this.viewEntityZ));
+    }
 }
