--- ../src-base/minecraft/net/minecraft/client/gui/GuiRepair.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiRepair.java
@@ -93,8 +93,8 @@
 
                 if (this.field_146289_q.func_82883_a())
                 {
-                    func_73734_a(k - 3, l - 2, this.field_146999_f - 7, l + 10, -16777216);
-                    func_73734_a(k - 2, l - 1, this.field_146999_f - 8, l + 9, -12895429);
+                	drawRectangle(k - 3, l - 2, this.field_146999_f - 7, l + 10, -16777216);
+                	drawRectangle(k - 2, l - 1, this.field_146999_f - 8, l + 9, -12895429);
                 }
                 else
                 {
